subxt/src/tx/tx_client.rs

```diff
@@ -445,8 +445,7 @@ where
     /// and obtain details about it, once it has made it into a block.
     pub async fn submit_and_watch(&self) -> Result<TxProgress<T, C>, Error> {
         // Get a hash of the extrinsic (we'll need this later).
-        let ext_hash = T::Hasher::hash_of(&self.encoded);
-
+        let ext_hash = T::Hasher::hash_of(&self.encoded.encode());
```

Comment:
can you explain why this is encoded twice?

Reply:
I'll add a comment. I think it is like this: The `self.encoded` refers to the raw bytes of the extrinsic. Calling the `.encode()` on these bytes basically just prefixes it with the compact encoded length, IIRC.

Reply:
I would rather add two separate fields `raw_bytes` and `scale_bytes` in both `SubmittableExtrinsic` and `ExtrinsicDetails` to avoid this `encode()`

These bytes are already allocated and can be reused later and it makes the code easier to understand.

Regarding https://github.com/paritytech/subxt/blob/master/subxt/src/rpc/rpc.rs#L401, that is encoding it to `sp_core::Bytes` so it may work just to pass the `raw_bytes` into there...

Reply:
I also thought about that, but it will not save us allocations, because inside of the `Rpc::watch_extrinsic()` function, we need to construct an `RpcParams` struct that needs to take ownership of the scale encoded bytes. So we need to clone them, because we cannot move them out of the SubmittableExtrinsic. What do you think about having the  `Rpc::watch_extrinsic()` function return the hash of the bytes of the `RpcParams`? This would avoid any unnecessary allocations. I think I need to talk with James about this issue on Monday, I also don't know if the fix I did is desired in this way.

Reply:
I don't care that much of the allocations to be honest but it would make the code much more readable.

The `rpc_params!` doesn't need to take ownership of the bytes FWIW, it just serializes them. 
So you could do `rpc_params![&data]`